doc/api/esm.md

```diff
@@ -1096,12 +1096,132 @@ export async function resolve(specifier,
 With this loader, running:
 
 ```console
-NODE_OPTIONS='--experimental-loader ./custom-loader.mjs' node x.js
+NODE_OPTIONS='--experimental-loader ./custom-loader.mjs' node main.js
 ```
 
-would load the module `x.js` as an ES module with relative resolution support
+would load the module `main.js` as an ES module with relative resolution support
 (with `node_modules` loading skipped in this example).
 
+### Get source hook
+
+The `getSource` hook provides a way to define a custom method for retrieving
+the source code of an ES module specifier. This would allow a loader to
+potentially avoid reading files from disk.
+
+```js
+import { get } from 'https';
+
+/**
+ * @param {object} module
+ * @param {string} module.url
+ * @param {string} module.format
+ * @param {function} defaultGetSource
+ * @returns {string|Promise}
+ */
+export async function getSource({ url, format }, defaultGetSource) {
+  if (!url.startsWith('https://')) {
+    return defaultGetSource(url);
+  } else {
+    return new Promise((resolve, reject) => {
+      get(url, (res) => {
+        let data = '';
+        res.on('data', (chunk) => data += chunk);
+        res.on('end', () => resolve(data));
+      }).on('error', (err) => reject(err));
+    });
+  }
+}
+
+export async function resolve(specifier,
+                              parentModuleURL,
+                              defaultResolver) {
+  if (specifier.startsWith('https://')) {
+    return {
+      url: specifier,
+      format: 'module'
+    };
+  } else if (parentModuleURL &&
+    parentModuleURL.startsWith('https://')) {
+    return {
+      url: new URL(specifier, parentModuleURL).href,
+      format: 'module'
+    };
+  } else {
+    return defaultResolver(specifier, parentModuleURL);
+  }
+}
+```
+
+With this loader, running:
+
+```console
+NODE_OPTIONS='--experimental-loader ./custom-loader.mjs' node main.js
+```
+
+Will enable `main.js` to contain a statement like
+`import { foo } from 'https://example.com/es-module.js'`, which normally isn’t
+supported by Node.js as the specifier isn’t a `file://` or `data://` URL.
+
+### Transform source hook
+
+The `transformSource` hook provides a way to modify the source code of a loaded
+ES module file after the source string has been loaded from disk but before
+Node.js has done anything with it.
+
+If this hook is used to convert unknown-to-Node.js file types into executable
+JavaScript, a resolve hook is also necessary in order to register any
+unknown-to-Node.js file extensions.
+
+```js
+import CoffeeScript from 'coffeescript';
+
+/**
+ * @param {object} module
+ * @param {string} module.url
+ * @param {string} module.format
+ * @param {string} source
+ * @returns {Promise|string}
+ */
+export async function transformSource({ url, format }, source) {
```

Comment:
this should provide a list of acceptable formats that the runtime/current loader can accept (probably as an argument, but could a runtime call instead).

Also, to support WASM and to align with the current implementation, it should be returning a buffer not a string (some formats do coerce to string though/we could document that).

This also should have an example of `getSource` returning an unsupported format (e.g. `text/coffeescript`) and returning a supported format like ESM (`text/javascript`)